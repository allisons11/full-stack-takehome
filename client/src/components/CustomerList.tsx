import { useQuery } from "@apollo/client";
import CustomerSummary from "./CustomerSummary";
import { GET_CUSTOMERS } from "@/graphql/queries";
import styles from "../styles/CustomerList.module.css";

interface Customer {
  id: string;
  username: string;
  name: string;
  address: string;
}

const CustomerList = () => {
  const { loading, error, data } = useQuery(GET_CUSTOMERS);

  if (loading) return <div>Loading</div>;
  if (error) return <div>Error fetching customer data: {error.message}</div>;
  if (!data) return <div>No data available</div>;

  const { Customers } = data;

  const summaries = Customers.map((customer: Customer, i: number) => (
    <tr key={i}>
      <CustomerSummary {...customer} />
    </tr>
  ));

  return (
    <div className={styles.customerList}>
      <table className={styles.table}>
        <thead>
          <tr>
            <th className={styles.headerCell}>Name</th>
            <th className={styles.headerCell}>Username</th>
            <th className={styles.headerCell}>Address</th>
          </tr>
        </thead>
        <tbody>{summaries}</tbody>
      </table>
    </div>
  );
};

export default CustomerList;
