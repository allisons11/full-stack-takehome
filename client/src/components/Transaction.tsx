import styles from "../styles/Transaction.module.css";

export interface TransactionProps {
  code: String;
  date: Number;
  amount: Number;
  price: Number;
  total: Number;
}

const Transaction = ({
  code,
  date,
  amount,
  price,
  total,
}: TransactionProps) => {
  const readableDate = new Date(Number(date)).toUTCString();
  return (
    <div className={styles.transaction}>
      <div>Date: {readableDate}</div>
      <div>Type: {code}</div>
      <div>Amount: {amount.toString()}</div>
      <div>Price: {price.toString()}</div>
      <div>Total: {total.toString()}</div>
    </div>
  );
};

export default Transaction;
