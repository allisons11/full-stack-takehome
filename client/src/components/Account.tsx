import { useQuery } from "@apollo/client";
import Transaction from "./Transaction";
import { GET_TRANSACTIONS } from "../graphql/queries";
import styles from "../styles/Account.module.css";

export interface AccountProps {
  accountId: Number;
}

export interface Transaction {
  code: String;
  date: Number;
  amount: Number;
  price: Number;
  total: Number;
}

const Account = ({ accountId }: AccountProps) => {
  const { loading, error, data } = useQuery(GET_TRANSACTIONS, {
    variables: { accountId },
  });
  if (loading) return <div>Loading</div>;
  if (error) return <div>Error fetching account data: {error.message}</div>;
  if (!data) return <div>No data available</div>;
  const { TransactionsPerAccount } = data;

  const transactions = [...TransactionsPerAccount]
    .sort((a: Transaction, b: Transaction) => {
      return Number(a.date) - Number(b.date);
    })
    .map((transaction: Transaction, i: number) => {
      const { code, date, amount, price, total } = transaction;
      return (
        <Transaction
          key={i}
          code={code}
          date={date}
          amount={amount}
          price={price}
          total={total}
        ></Transaction>
      );
    });
  return (
    <article className={styles.accountCard}>
      <header>
        <h3>Account # {accountId.toString()}</h3>
      </header>
      <section>
        <h4>Transaction History</h4>
        {transactions}
      </section>
    </article>
  );
};

export default Account;
