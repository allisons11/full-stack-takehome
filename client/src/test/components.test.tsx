import React from "react";
import { render } from "@testing-library/react";
import {
  ApolloProvider,
  InMemoryCache,
  ApolloClient,
  HttpLink,
} from "@apollo/client";
import fetch from "cross-fetch";

import Account, { AccountProps } from "../components/Account";
import CustomerList from "@/components/CustomerList";
import Transaction, { TransactionProps } from "@/components/Transaction";

let client: any;

beforeAll(() => {
  client = new ApolloClient({
    link: new HttpLink({ uri: "http://localhost:4000", fetch }),
    cache: new InMemoryCache(),
  });
});

describe("component snapshot tests", () => {
  test("CustomerList", () => {
    const { container } = render(
      <ApolloProvider client={client}>
        <CustomerList />
      </ApolloProvider>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
  test("Account", () => {
    const props: AccountProps = { accountId: 774823 };
    const { container } = render(
      <ApolloProvider client={client}>
        <Account {...props} />
      </ApolloProvider>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
  test("Transaction", () => {
    const props: TransactionProps = {
      code: "buy",
      date: 1239309238,
      amount: 200.234234,
      price: 9.0924234,
      total: 99999,
    };
    const { container } = render(
      <ApolloProvider client={client}>
        <Transaction {...props} />
      </ApolloProvider>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
