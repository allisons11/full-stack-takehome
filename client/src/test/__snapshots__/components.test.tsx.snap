// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component snapshot tests Account 1`] = `
<div>
  Loading
</div>
`;

exports[`component snapshot tests CustomerList 1`] = `
<div>
  Loading
</div>
`;

exports[`component snapshot tests Transaction 1`] = `
<div
  class="transaction"
>
  <div>
    Date: 
    Thu, 15 Jan 1970 08:15:09 GMT
  </div>
  <div>
    Type: 
    buy
  </div>
  <div>
    Amount: 
    200.234234
  </div>
  <div>
    Price: 
    9.0924234
  </div>
  <div>
    Total: 
    99999
  </div>
</div>
`;
